### PR TITLE
[SPARK-37436][PYTHON] Uses Python's standard string formatter for SQL API in pandas API on Spark

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to use [Python's standard string formatter](https://docs.python.org/3/library/string.html#custom-string-formatting) instead of hacky custom SQL parser for SQL API in pandas API on Spark

### Why are the changes needed?

Current implementation of parsing is very hacky, and does not work. It is [dependent on Python's internal module](https://github.com/apache/spark/blob/master/python/pyspark/pandas/sql_processor.py#L291), and [Series is being treated as a table](https://github.com/apache/spark/blob/master/python/pyspark/pandas/sql_processor.py#L339-L340), etc.

We should have the Python standard string formatter with the standard interface and the standard support.

### Does this PR introduce _any_ user-facing change?

Yes.

**Disallowed:**

1. `Series` as a table

    ```python
    myser = ps.Series({'a': [1.0, 2.0, 3.0], 'b': [15.0, 30.0, 45.0]})
    ps.sql("SELECT * from {myser}", myser=myser)
    ```

2. `list` and `range`

    ```python
    strs = ['a', 'b']
    ps.sql("SELECT 'a' IN {strs}", strs=strs)
    ```

3. Automatic local/global variable detection:

    ```python
    strs = ['a', 'b']
    ps.sql("SELECT 'a' IN {strs}")
    ```

**Allowed:**

1. `Series` as a column

    ```python
    mydf = ps.range(10)
    ps.sql("SELECT {ser} FROM {mydf}", ser=mydf.id, mydf=mydf)
    ```

2. Reference checking (between `Series` and `DataFrame`)

    ```python
    mydf = ps.range(10)
    ps.sql("SELECT {ser} FROM tblA", ser=mydf.id)
    ```

    ```
    ValueError: The series in {ser} does not refer any dataframe specified.
    ```

3. Attribute supports from frame (standard Python support):

    ```python
    mydf = ps.range(10)
    ps.sql("SELECT {tbl.id} FROM {tbl}", tbl=mydf)
    ```

### How was this patch tested?

Doctests were added.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
This is ready for a review. cc @ueshin @xinrong-databricks @itholic.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

File: python/pyspark/pandas/sql_processor.py [hunk 1]
```diff
@@ -43,142 +43,6 @@ def sql(
     locals: Optional[Dict[str, Any]] = None,
     **kwargs: Any,
 ) -> DataFrame:
-    """
```
Comment (xinrong-databricks) [inline python/pyspark/pandas/sql_processor.py:46]: I am curious why the docstring is removed.

File: python/pyspark/pandas/sql_formatter.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    Also, it is possible to query using Series.
+
+    >>> psdf = ps.DataFrame({"A": [1, 2, 3], "B":[4, 5, 6]}, index=['a', 'b', 'c'])
+    >>> ps.sql("SELECT {mydf.A} FROM {mydf}", mydf=psdf)
+       A
+    0  1
+    1  2
+    2  3
+    """
+    if "PYSPARK_PANDAS_SQL_LEGACY" in os.environ:
```
Comment (xinrong-databricks) [inline python/pyspark/pandas/sql_formatter.py:155]: Do we intend not to check `PYSPARK_PANDAS_SQL_LEGACY` equals to 1 or not?

**Review by xinrong-meng — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**